packages/core/src/declareAtom.ts

```diff
@@ -51,167 +76,123 @@ export function declareAtom<TState>(
-    dep: Unit | PayloadActionCreator<T>,
-    reducer: Reducer<TState, T>,
-  ) {
+  function on<T>(dep: Unit<any>, reducer: Reducer<TState, T>) {
     if (!initialPhase)
-      throwError("Can't define dependencies after atom initialization")
-
-    safetyFunc(reducer, 'reducer')
+      throwError(`"on" handler calls after initialization in atom "${selfId}"`)
 
     const position = dependencePosition++
-    const depTree = getTree(dep as Unit)!
-    if (!depTree) throwError('Invalid dependency')
-    const depId = depTree.id
 
+    safetyFunc(reducer, 'atom reducer №' + position)
+
+    const depNode = getNode(dep)
+    if (!depNode) throwError('atom dependency №' + position)
+    const depId = depNode.id
     const isDepActionCreator = getIsAction(dep)
 
-    _tree.union(depTree)
+    depNode.depsAll.forEach(
+      id => !selfDepsAll.includes(id) && selfDepsAll.push(id),
+    )
+    selfDepsAll.push(depId)
+    selfDeps.push(depNode)
+
+    selfUpdates.push(({ state, stateNew, payload, changedIds, type }: Ctx) => {
+      if (!depNode.depsAll.includes(type)) return
 
-    function update({ state, stateNew, payload, changedIds, type }: Ctx) {
-      const atomStateSnapshot = state[_id]
-      // first `walk` of lazy (dynamically added by subscription) atom
-      const isAtomLazy = atomStateSnapshot === undefined
+      const atomStateSnapshot = state[selfId]
+      const isFirstWalk = atomStateSnapshot === undefined
 
-      if (!isAtomLazy && type === initAction.type) return
+      if (!isFirstWalk && type === initAction.type) return
 
-      const atomStatePreviousReducer = stateNew[_id]
+      const atomStatePreviousReducer = stateNew[selfId]
       // it is mean atom has more than one dependencies
       // that depended from dispatched action
       // and one of the atom reducers already processed
       const hasAtomNewState = atomStatePreviousReducer !== undefined
-      const atomState = hasAtomNewState
+      const atomState = (hasAtomNewState
         ? atomStatePreviousReducer
-        : atomStateSnapshot
+        : atomStateSnapshot) as TState
 
       const depStateSnapshot = state[depId]
       const depStateNew = stateNew[depId]
       const isDepChanged = depStateNew !== undefined
       const depState = isDepChanged ? depStateNew : depStateSnapshot
-      const depValue = isDepActionCreator ? payload : depState
+      const depValue = (isDepActionCreator ? payload : depState) as T
 
-      if (isDepActionCreator || isDepChanged || isAtomLazy) {
+      if (atomState === undefined && type !== initAction.type)
+        throwError('dispatch with undefined state')
+      if (isDepActionCreator || isDepChanged || isFirstWalk) {
         const atomStateNew = reducer(atomState, depValue)
 
         if (atomStateNew === undefined)
           throwError(
-            `Invalid state. Reducer № ${position} in "${_id}" atom returns undefined`,
+            `(undefined) state in reducer № ${position} in atom "${selfId}"`,
```

Comment:
https://github.com/artalar/reatom/pull/213/files#diff-061a8306761a19f5a5a8453d3d63a652R82
> `initial state (undefined) in atom "${selfId}"`

but here
> `(undefined) state in reducer № ${position} in atom "${selfId}"`

(order in message)